src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/DoNotGuardDictionaryRemoveByContainsKey.cs

```diff
@@ -0,0 +1,155 @@
+            CreateLocalizableResourceString(nameof(DoNotGuardDictionaryRemoveByContainsKeyDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false,
+            additionalCustomTags: WellKnownDiagnosticTags.Unnecessary);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private static void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!TryGetDictionaryTypeAndContainsKeyMethod(context.Compilation, out var dictionaryType, out var containsKeyMethod))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(context => AnalyzeOperation(context, dictionaryType, containsKeyMethod), OperationKind.Conditional);
+
+            static void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol dictionaryType, IMethodSymbol containsKeyMethod)
+            {
+                var conditionalOperation = (IConditionalOperation)context.Operation;
+
+                IInvocationOperation? invocationOperation = null;
+
+                switch (conditionalOperation.Condition)
+                {
+                    case IInvocationOperation iOperation:
+                        invocationOperation = iOperation;
+                        break;
+                    case IUnaryOperation unaryOperation when unaryOperation.OperatorKind == UnaryOperatorKind.Not:
+                        if (unaryOperation.Operand is IInvocationOperation operand)
+                            invocationOperation = operand;
+                        break;
+                    default:
+                        return;
+                }
+
+                if (invocationOperation == null || !invocationOperation.TargetMethod.OriginalDefinition.Equals(containsKeyMethod, SymbolEqualityComparer.Default))
+                {
+                    return;
+                }
+
+                if (conditionalOperation.WhenTrue.Children.Any())
+                {
+                    var properties = ImmutableDictionary.CreateBuilder<string, string?>();
+                    properties[ConditionalOperation] = CreateLocationInfo(conditionalOperation.Syntax);
+
+                    switch (conditionalOperation.WhenTrue.Children.First())
+                    {
+                        case IInvocationOperation childInvocationOperation:
+                            if (childInvocationOperation.TargetMethod.Name == "Remove" &&
+                                childInvocationOperation.TargetMethod.OriginalDefinition.ContainingType.Equals(dictionaryType, SymbolEqualityComparer.Default))
```

Comment:
Why are we checking the method name and then comparing the symbol for containing type? Instead we should adjust `TryGetDictionaryTypeAndContainsKeyMethod` to also compute the method symbol for the `Remove` method on the dictionary type and probably rename the method to `TryGetDictionaryContainsKeyAndRemoveMethods`?

Reply:
I guess it is for covering both overloads of `Dictionary.Remove(...)`, though probably better to use symbol compare even there were 2 overloads needs to be flagged

Reply:
Yes, we should always compare symbols, not strings. Can you please change this to store all the symbol overloads and compare with them?

Reply:
Updated